### ISSUE TITLE
Potential memory leak

### ISSUE THREAD

**Issue by anhldbk** [mcollina/mosca#479]
Thank all you guys for this great package.

I'm performing some kind of stress testing on `Mosca`. The scenario is pretty simple. A client just sends many message to `Mosca`, and I will measure how `Mosca` utilizes CPU cores, memory...

The code is below:

**Server**
```javascript
var mosca = require('mosca');
var settings = {
  port: 1884,
  backend: {
    type: 'redis',
    redis: require('redis'),
    db: 12,
    port: 6379,
    host: "localhost"
  }
};
var server = new mosca.Server(settings);

server.on('ready', function(){
  console.log('Mosca server is up and running');  
});
```

**Client**
```javascript
var mqtt    = require('mqtt');
var endpoint = mqtt.connect({ host: 'localhost', port: 1884 });

endpoint.on('connect', function () {
  endpoint.subscribe('/topic');

  setTimeout(function(){
    var count = 0;
    while(count++ < 100000000){
      endpoint.publish('/topic', 'tada'+count);
    }
  }, 1000);
});

endpoint.on('message', function (topic, message) {
  console.log(message.toString());
});
```

There are several things worth to mention:
+ The server's memory usage gradually increases
+ If I terminate the client, the memory usage above stays the same

Sometimes, I've got some nasty exceptions:
**Server**
![screenshot from 2016-05-19 22-21-45](https://cloud.githubusercontent.com/assets/3270746/15399118/9e15d786-1e10-11e6-84d3-250027d68369.png)

**Client**
![screenshot from 2016-05-19 22-21-32](https://cloud.githubusercontent.com/assets/3270746/15399121/a10af08e-1e10-11e6-9ecb-9f5894ae9f1c.png)

Would you pls tell me what wrong is?

**Comment by mcollina** [comment]
1. I'm not sure about the HiRedis error, probably you are using an unsupported Redis version, or something similar, I've never seen it. It's definitely not a Mosca issue (as I know of)
2. the *client* is running out of memory, because you are not using backpressure to send all those data. send messages via the callback of `publish`, and not in a while loop. In your test, you are basically allocating 100000000 packets and then trying to send them all at the same time.

Regarding the gc, I need some code and more example. My experience is that memory consumption is really high, but it does not leak.

If you want more a more performant implementation, check https://github.com/mcollina/aedes. I'll start using mosca on top of that after mosca v2 is out.

**Comment by anhldbk** [comment]
@mcollina So you recommend me to modify the client to be like:

```javascript
endpoint.on('connect', function () {
  console.log('connected');
  endpoint.subscribe('/topic');

  setTimeout(function(){
      var count = 0;
      var fn = function(){
          endpoint.publish('/topic', 'tada'+ count, callback=function(){
            console.log(count)
            count += 1
            fn()
          })
      }
      fn()
  }, 1000);
});
```
Btw, I'm using `redis` v3.2 and my `package.json` has `redis": "^2.6.0-2"`

**Comment by anhldbk** [comment]
@mcollina Now I know how to work with [back-pressure](https://github.com/mcollina/mqtt-level-store). 
Still I'm not sure what wrong with the server is.

**Comment by mcollina** [comment]
@anhldbk does that happen all the time? Or in some weird condition?

cc @behrad which version of Redis are you running?

**Comment by behrad** [comment]
> cc @behrad which version of Redis are you running?

I'm currently on my fork of Mosca based on ioredis `1.5` with redis server `3.0.7`. My older setup was on default Mosca with current master's version of `node_redis` without hiredis.

**Comment by anhldbk** [comment]
@mcollina Yep, it happens all the time.

**Comment by mcollina** [comment]
Probably a duplicate of https://github.com/mcollina/mosca/issues/476

**Comment by mcollina** [comment]
Probably fixed in v2.0.0. Feel free to reopen.
